### ISSUE TITLE
Convert `id_method_name` to `foreign_key`

### ISSUE THREAD

**Issue by christophersansone** [fast-jsonapi/fast_jsonapi#45]
The option `id_method_name` is confusing, and it is currently referring to two different things:

1. When `polymorphic` is not specified, then `id_method_name` is the property name for the _foreign key_ lookup, e.g. `user.document_ids`.

2. When `polymorphic` is specified, then `id_method_name` is assumed to be the _primary key_ of the related object, e.g. `user.documents.map(&:id)`.

Let's consider changing `id_method_name` to simply `foreign_key`.  This is similar to ActiveRecord, and it is more clear that it should refer to the foreign key rather than the primary key.

```ruby
class UserSerializer

  # the `foreign_key` becomes `company_id` by default
  belongs_to :company

  # the `foreign_key` becomes `document_ids` by default
  has_many :documents

end
```

So `has_many` relationships are automatically pluralized by default.  And of course, the default value can be overridden with `options[:foreign_key]`.

The library also needs some clear logic to understand when the foreign key can be used vs. when it needs to read the related object(s).  At first glance, it seems like it should read the related objects under the following conditions:

* The serializer is dynamic (i.e. polymorphic or a Proc), and `options[:foreign_key]` is not specified.
* The relationship is defined with a block, and `options[:foreign_key]` is not specified.
* The record does not respond to the property, e.g. `user.respond_to?('document_ids')

In all other cases, it should use the foreign key.

Lastly, when the serializer does need to read the related object(s), it should determine its ID property name using its serializer's `set_id` value.  It should not attempt to use the foreign key.

I believe we can effectively deprecate `id_method_name` and use this methodology instead.

**Comment by stas** [comment]
@christophersansone I'm looking at the JSON:API spec, and I don't see how the proposed naming is a better alternative:
https://jsonapi.org/format/#document-resource-object-identification

I also don't think an `id` always represents a foreign-key.

I believe we can just document properly when and how the `id_method_name` is used if that's the main source of confusion.

**Comment by christophersansone** [comment]
@stas Thanks for the quick feedback!  I didn't do a great job of explaining the actual problem, and I jumped right to the solution.  There are actually a few different issues worth discussing in detail.  Let me explain with some examples.

Before I begin, just a quick clarification to your comment:

> I also don't think an id always represents a foreign-key.

Correct that it is not necessarily a foreign key constraint at the database level, but the idea is the same: a property of the record is referencing one or many unique identifiers of different records.  It could be called `foreign_key` or `foreign_id` but the point is that it _solely_ used for this purpose, which is not the case for `id_method_name`.

Before we can understand the desire to change the naming convention, let's first take a look at how `id_method_name` currently works.  The basic usage is as follows:

```
class Zoo < ActiveRecord::Base
  has_many :employees
end

class Employee < ActiveRecord::Base
  belongs_to :zoo
end

class ZooSerializer
  include FastJsonapi::ObjectSerializer
  has_many :employees
end

class EmployeeSerializer
  include FastJsonapi::ObjectSerializer
  belongs_to :zoo
end

zoo = Zoo.first
ZooSerializer.new(zoo).serializable_hash
```

Example output:

```
{
  data: {
    type: 'zoo',
    id: '1',
    attributes: { name: 'Lincoln Park Zoo' },
    relationships: {
      employees: [
        { type: 'employee', id: '123' },
        { type: 'employee', id: '456' }
      ]
    }
  }
}
```

In this scenario, the serialization only needs the ID for the employees, not the full record.  There are two general ways to accomplish this (at least in ActiveRecord): `zoo.employees.map(&:id)` and `zoo.employee_ids`.  The first queries the full employee records from the database, builds the models, and iterates over them to return the list of IDs.  The second is a shortcut to "pluck" only the IDs from the database if the records are not already loaded, or iterate over the model if already loaded.  The second option is preferred for performance.  This is where `id_method_name` comes in.

By default, a serializer's relationship will try to determine the name of this ID method and store it in `id_method_name`.  In this case, the `has_many :employees` correctly infers the property name as `employee_ids`.  This inferred value can be overridden by manually specifying `options[:id_method_name]`.

If this is the extent of the use cases, it's perfect.  I imagine this is how it was initially designed.  However, there is more to the story...

One fundamental problem is this.  What happens if we are not using ActiveRecord, for example maybe we are just serializing some general classes:

```
Employee = Struct.new(:id)
Zoo = Struct.new(:id, :employees)

employee1 = Employee.new('123')
employee2 = Employee.new('456')
zoo = Zoo.new('1', [ employee1, employee2 ])

ZooSerializer.new(zoo).serializable_hash
```

This should work, right?  Well it does not.  It throws an exception because the serializer is expecting `zoo.employee_ids` to exist.  We can easily just say "well it should check if `zoo.respond_to?(:employee_ids)` first", which is probably correct... but then what should happen in this case?

Today, how does a developer address this case?  Would defining the relationship as a block help?  I don't know the answer here.  And I have run into this a couple times.  I wind up just writing an `employee_ids` method myself on the parent class just to get it working.

That is one problem, but it is also probably the simplest issue.  Consider this:

```ruby
class ZooSerializer
  ...
  has_many :animals, polymorphic: true
end
```

What happens in this case?  Does it call `zoo.animal_ids`?  Currently it does not.  Currently, it iterates over the objects, i.e. `zoo.animals.map(&:id)`.  This is where `id_method_name` gets extremely confusing.  The `polymorphic` options changes its behavior from the foreign key `zoo.animal_ids` to the primary key `animal.id`.  This behavior change also occurs when the relationship is defined as a block, and also (once #32 is merged) when the serializer is a Proc... in other words, any time that the serializer is not _static_.

My concerns about this change in behavior:

* The behavior change is hard-coded to these situations and not configurable by the developer.  Some situations use `zoo.animal_ids` and some use `zoo.animals.map(&:id)`.
* It is unclear to the developer which one is occurring at any time.  And honestly, the internal behavior may change if we find better ways to optimize.  Which one occurs should not be a concern of the developer, unless they have somehow indicated in `options` to do it one way or the other.
* If the behavior is unclear in certain situations, then how will a developer know to set `id_method_name` to the foreign key (`animal_ids`) or the primary key (`animal.id`)?
* Currently, when the serializer iterates over the objects, it is hard-coded to use `id` as the primary key when `options[:id_method_name]` is not specified.

All this said, from a DSL standpoint, it is extremely unclear in what cases someone should override `id_method_name`, and if so, what that value should be.

Now, here is something that will help the situation.  Once #32 is merged, we will have a much better way to determine the primary key: look up the serializer itself and use `set_id`.  This way, an object's serializer becomes the Single Source of Truth to determine both the `type` and `id`.  Much like the `record_type` change, there would be no reason to specify `id_method_name` in this situation.  And therefore, specifying the "ID method" would only be needed in the foreign key situation.

Therefore, we can make `id_method_name` exclusive to the foreign key lookup.  And if we do that, I thought we can change the name to make it clear.  We don't need to, it can stay `id_method_name`, but I figured that `foreign_key` could give the DSL a fresh start.

**Comment by christophersansone** [comment]
Here is the pseudocode of the logic as I see it:

```ruby
class Relationship
  ...
  def serialize(record, params)
    if static_serializer && record.respond_to?(foreign_key)
      # use the foreign key
      foreign_ids = record.send(foreign_key)
      if has_many
        return foreign_ids.map { |id| { type: static_serializer.record_type, id: id } }
      else
        { type: static_serializer.record_type, id: id }
      end
    else
      # use the dynamic / primary key lookup
      if has_many
        objects = record.send(object_method_name)
        return objects.map do |object|
          serializer = serializer_for(object, params)
          { type: serializer.record_type, id: serializer.id_for(object, params) }
        end
      else
        object = record.send(object_method_name)
        serializer = serializer_for(object, params)
        { type: serializer.record_type, id: serializer.id_for(object, params) }
      end
    end
  end
end
```

It only uses `foreign_key` if it is a static serializer and the method exists on the record.  If the serializer is dynamic, then the record type may also be dynamic, and therefore we need to iterate over the objects to get the record type and therefore might as well get the ID as well.

This way, `foreign_key` just becomes a shortcut that can be used if the serializer determines that it can.  And if the serializer determines that it needs to iterate over the objects for whatever reason, it will disregard `foreign_key` and just get the record type and ID property from the object's serializer.

To me, this is a much clearer approach.  It is easier to reason about and easier to communicate and document.

**Comment by stas** [comment]
@christophersansone I agree that the serializer `id_method_name` (be that a single record, or a relationship serializer) should be used exclusively to take care of the identifier value, though I still don't see how this is relevant to the `foreign_key` naming.

Finally, I'd like to remind us all that this library should be agnostic towards what are the objects it is used to serialize. So to me, `foreign_key` is inferior to the `id_method_name`, simply because we don't have to explain what `id_method_name` is, it is already covered by the JSON:API spec. Otherwise we'll end up with another #23 :cold_sweat: 

One important aspect here though, is that we need to take care of the performance _penalty_ after such changes, since there's a high chance that was the main reason why this design decision was considered in the first place.

**Comment by christophersansone** [comment]
@stas I don't think there is anything intrinsically different here that would cause a performance penalty.  Most times the method name will be memoized up front.

I am fine with keeping it `id_method_name` if you feel that strongly about it.  I am more concerned with the fact that it currently can mean two different things, based on the other options specified.  My thought to change the naming convention is more about _ensuring it is only used one way_, which of course we can do with out changing the name.

**Comment by aguynamedben** [comment]
+1 to this

I'm new to jsonapi-serializer, and it took a while to get this right...

```rb
class Thing
  include JSONAPI::Serializer
  belongs_to :created_by, id_method_name: :created_by_uuid, record_type: :user, serializer: :organization_peer
  # what I tried first...
  belongs_to :created_by, foreign_key: :created_by_uuid, record_type: :user, serializer: :organization_peer
end
```

My relationship in ActiveRecord
```rb
class Thing < ApplicationRecord
  belongs_to :created_by, class_name: 'User', foreign_key: :created_by_uuid, primary_key: :uuid
end
```

**Comment by stas** [comment]
This is now handled in #141 

The `id_method_name` stays, but it will be used consistently in the relationships. This way there's no other way to define a method name to fetch the ID. The `ids_method_name` is available to define the method for IDs for relationships.
